### ISSUE TITLE
Trajectories should be deque, not lists (while generating)

### ISSUE THREAD

**Issue by sroet** [openpathsampling/openpathsampling#1099]
When looking over #1066 I noticed that when the engine is running backwards we are adding to the trajectory via:
```
trajectory.insert(0, snapshot.reversed)
```
This is slow, and becomes slower when the list is longer. As we are only ever appending to the trajectory on the ends, the appropriate data structure should be `deque` instead. 
![Screenshot from 2022-01-17 11-18-10](https://user-images.githubusercontent.com/19587267/149752327-f604f4ba-1a05-4eae-9206-4bdca1969d63.png)

Access by index becomes slower (especially in the middle) so that might need to be handled. Ideally a trajectory would be a `deque` while running/generating and converting them to a `list` when finalizing.
![Screenshot from 2022-01-17 11-32-50](https://user-images.githubusercontent.com/19587267/149753808-53b262d5-5136-4c61-872a-bbe578694d88.png)

**Comment by sroet** [comment]
Here are some runs for the 1000-10_000 range (more typical for trajectories)
## 1000

![Screenshot from 2022-01-17 11-37-07](https://user-images.githubusercontent.com/19587267/149754491-bc3bf3b5-cdf9-4a2a-95ed-aba7ae34379f.png)

## 10_000

![Screenshot from 2022-01-17 11-37-58](https://user-images.githubusercontent.com/19587267/149754503-cf869cee-4c03-489b-92fd-878d7109d55a.png)

**Comment by dwhswenson** [comment]
True, but is there any reason to believe that this is a performance bottleneck in real-world use? I think that we have enough overhead in other places that we'd get better performance improvements elsewhere.

An update for this would be welcome, but I don't think it's urgently required (so I probably won't implement it, personally).

One note: some aspects around the `DynamicsEngine.generate` and `DynamicsEngine.iter_generate` are likely to be significantly changed in 2.0. First part of that change in #1042, but I'm also don't particularly like the way `DynamicsEngine.generate` does this `for trajectory in it: pass` thing -- it seems wasteful to create an entire new trajetory object that we just throw away over and over again (overhead of UUIDs, etc). I'm open to significant changes in that for OPS 2.0, such as returning an iterator of snapshots instead of trajectories and only wrapping it in a trajectory in the `generate` method.

**Comment by sroet** [comment]
> True, but is there any reason to believe that this is a performance bottleneck in real-world use? I think that we have enough overhead in other places that we'd get better performance improvements elsewhere.

I would guess so as well (outside of strange toy-engines that are both fast and generate long trajectories). This was just a thing I noticed when looking over that PR and I made this issue to get it off my (mental) list. I agree that this should not have any priority.
